src/lib/operators/join_hash.cpp

```diff
@@ -215,8 +221,15 @@ class JoinHash::JoinHashImpl : public AbstractJoinOperatorImpl {
           ChunkOffset offset = 0;
           for (auto&& elem : materialized_chunk) {
             if (elem.first.chunk_offset != INVALID_CHUNK_OFFSET) {
-              output[row_id] =
-                  PartitionedElement<T>{RowID{chunk_id, offset}, murmur2<T>(elem.second, seed), elem.second};
+              uint32_t hashed_value;
+
+              if constexpr (JoinHashTraits<LeftType, RightType>::needs_lexical_cast && !std::is_same_v<T, HashType>) {
+                hashed_value = murmur2<HashType>(boost::lexical_cast<HashType>(elem.second), seed);
```

Comment:
Any reason not to use `type_cast` here? The fact that we use `boost::lexical_cast` (or any type of lexical cast) should be hidden in a central file that deals with type conversions, currently `type_cast.hpp`.

Reply:
`type_cast` is for `AllTypeVariant` values, not integral types. And I would disagree with your point about hiding these type conversions because I think that these are conversions specific to the Hash Join

Reply:
nevermind, I think `type_cast` works. I'm guessing that the values are implicitly cast to `AllTypeVariant`

Reply:
So using `type_cast` seems to work, and I wouldnt even need the `JoinHashTraits` anymore. However, this adds a lot of overhead. First of all, every value will be implicitly wrapped in an `AllTypeVariant` before being type-cast. That means unnecessary wrapping and unwrapping in cases where no real cast is necessary. Also, `type_cast` is not a `constexpr` method like it would be with the `JoinHashTraits`. 

I implemented the `type_cast` version in the current commit 2b290eb. Please let me know what you think

Reply:
This seems to be a problem with `type_cast`. We should have some `type_cast` version that takes two data types, checks if they are convertible and only uses `lexical_cast` if they are not.

Reply:
moved to #718